### PR TITLE
scripts: add option to generate junit xml reports

### PR DESCRIPTION
**Pull request by Rajalakshmi-Girish**
scripts: add option to generate junit xml reports

To add the option of generating junit xml test reports using the gotest tool `gotestsum`.
This modifies the script `test_lib.sh` to check for the variable `JUNIT_REPORT_DIR`/ `ARTIFACTS` and thus creating a folder and subsequent stdout and xml files.
The go flags -v and -json are enabled for the `go test` command when JUNIT report is requested.

### CODE REVIEW

**Comment by mkumatag** [comment]
someone can help us review this PR? this will be beneficial for any modern tools to report the tests in a nice format @spzala @gyuho

**Comment by lilic** [comment]
Hey, thank you for the PR, can you maybe explain or add some docs for this? Is this more useful for local test running or CI?

**Comment by Rajalakshmi-Girish** [comment]
> 
> 
> Hey, thank you for the PR, can you maybe explain or add some docs for this? Is this more useful for local test running or CI?

This change will be more useful for CI tools to have the test reports generated in junit xml format which is easy to read and analyze.
Currently, the tests run against `etcd` repo display the `go test` output as it is in the logs and would be difficult to go through and figure out the count of failures and the root causes. This Junit report generation has greater readability.

This change is inspired by https://github.com/kubernetes/kubernetes/blob/master/hack/make-rules/test.sh#L221:L259

**Comment by lilic** [comment]
> This change will be more useful for CI tools to have the test reports generated in junit xml format which is easy to read and analyze.

Thanks for the explanation! Do you mind including the link to this output that we can see in our CI here.

**Comment by mkumatag** [comment]
> > This change will be more useful for CI tools to have the test reports generated in junit xml format which is easy to read and analyze.
> 
> Thanks for the explanation! Do you mind including the link to this output that we can see in our CI here.

I see etcd project is using the GitHub actions for the CI, you might need to enhance the action workflow to display the junit tests, e.g: https://github.com/marketplace/actions/publish-unit-test-results which can be a subsequent change needed.

if you are interested in how it looks in our CI setup(prow based), here it is: https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-master-golang-etcd-build-test-ppc64le/1404900977844162560

**Comment by lilic** [comment]
> I see etcd project is using the GitHub actions for the CI, you might need to enhance the action workflow to display the junit tests, e.g: https://github.com/marketplace/actions/publish-unit-test-results which can be a subsequent change needed.

Thanks for that info! @ptabor @serathius for further thoughts if that is needed right now.

**Comment by Rajalakshmi-Girish** [comment]
@lilic is there any update on this please?

File: scripts/test_lib.sh [hunk 1]
```diff
@@ -236,11 +269,33 @@ function go_test {
...
+  fi
+
+  # Used to filter verbose test output.
+  go_test_grep_pattern=".*"
+
+  if [[ -n "${JUNIT_REPORT_DIR}" ]] ; then
+    goTestFlags+="-v "
+    goTestFlags+="-json "
+    # Show only summary lines by matching lines like "status package/test"
+    go_test_grep_pattern="^[^[:space:]]\+[[:space:]]\+[^[:space:]]\+/[^[[:space:]]\+"
```
Comment (ptabor) [inline scripts/test_lib.sh:294]: Does it mean that if the test fails, we will still see only summary without integration with the 'xml' viewer ?

Comment (ptabor) [reply scripts/test_lib.sh:294]: Could you, please, generate a run with artificially failing test to take a look how the results (logs) from the failed test-case would look.

Comment (Rajalakshmi-Girish) [reply scripts/test_lib.sh:294]: @ptabor The `go_test_grep_pattern` is to filter the `go test` output in the usual log/console and display only the Summary lines from `json` output when JUNIT Reporting is enabled.
When JUNIT Reporting is NOT enabled, the `go_test_grep_pattern` is ".*" thus displaying the complete output of `go test`.
Due to this `go_test_grep_pattern` when JUNIT is enabled, we get below kind of output:
image

Comment (Rajalakshmi-Girish) [reply scripts/test_lib.sh:294]: > 
> 
> Could you, please, generate a run with artificially failing test to take a look how the results (logs) from the failed test-case would look.

We have a failing test on our ppc64le architecture. Can you please tell me if the below link is accessible?
https://prow.ppc64le-cloud.org/view/s3/prow-logs/logs/postsubmit-master-golang-etcd-build-test-ppc64le/1404900977844162560

Comment (ptabor) [reply scripts/test_lib.sh:294]: thx. Seems good.

**Review by ptabor — APPROVED** [review]

**Comment by serathius** [comment]
I think that having junit reports will, provide much better visibility in tests failures. Instead of bing forced to track flakes one by one, with junit we should be able to identify them in bulk and start filing issues to bring attention of more people. At some point even automate the process.

**Review by serathius — APPROVED** [review]

**Comment by serathius** [comment]
@Rajalakshmi-Girish Are you planning to introduce integration with https://github.com/marketplace/actions/publish-unit-test-results? 

I'm happy to push continue this so we can get rid of those pesky flakes.

**Comment by serathius** [comment]
Sent https://github.com/etcd-io/etcd/pull/13152

**Comment by Rajalakshmi-Girish** [comment]
@serathius https://github.com/etcd-io/etcd/pull/13154 to fix a flaw in this PR.
Can you please take a look.

**Pull request merged**